result.nim

```diff
@@ -0,0 +1,352 @@
+  ## Would be nice if it was automatic...
+  if self.isOk: result.ok(self.value)
+  else: result.err(self.error)
+
+proc mapCast*[T0, E0](self: Result[T0, E0], T1: typedesc): Result[T1, E0] =
+  ## Convert result value to A using a cast
+  ## Would be nice with nicer syntax...
+  if self.isOk: result.ok(cast[T1](self.value))
+  else: result.err(self.error)
+
+template `and`*[T, E](self: Result[T, E], other: untyped): untyped =
+  ## Evaluate `other` iff self.isOk, else return error
+  ## fail-fast - will not evaluate other if a is an error
+  if self.isOk:
+    other
+  else:
+    type R = type(other)
+    R.err(self.error)
+
+template `or`*[T, E](self: Result[T, E], other: untyped): untyped =
+  ## Evaluate `other` iff not self.isOk, else return self
+  ## fail-fast - will not evaluate other if a is a value
+  if self.isOk: self
+  else: other
+
+template `+`*[T, E](self, other: Result[T, E]): untyped =
+  ## Perform `+` on the values of self and other, if both are
+  type R = type(other)
+  if self.isOk:
+    if other.isOk:
+      R.ok(self.value + other.value)
+    else:
+      R.err(other.error)
+  else:
+    R.err(self.error)
+
+template catch*(body: typed): Result[type(body), ref Exception] =
+  ## Convert a try expression into a Result
+  type R = Result[type(body), ref Exception]
+
+  try:
+    R.ok(body)
+  except:
+    R.err(getCurrentException())
+
+template capture*(a: typedesc, e: ref Exception): Result[a, ref Exception] =
+  type R = Result[a, ref Exception]
+
+  var ret: R
+  try:
+    # TODO is this needed? I think so, in order to grab a call stack, but
+    #      haven't actually tested...
+    if true:
+      # I'm sure there's a nicer way - this just works :)
+      raise e
+  except:
+    ret = R.err(getCurrentException())
+  ret
+
+proc `==`[E, T](lhs, rhs: Result[E, T]): bool =
+  if lhs.isOk != rhs.isOk: return false
+  if lhs.isOk: return lhs.value == rhs.value
+
+  return lhs.error == rhs.error
+
+proc `[]`*[T, E](self: Result[T, E]): T =
+  ## Fetch value of result if set, or raise error as an Exception
+  if self.isErr:
+    var e: ref ResultError
+    when compiles($e.error):
+      # Set exception message, iff we can convert error to a string
+      new (e, $e.error)
+    else:
+      new (e)
+    raise e
+  self.value
+
+proc `[]`*[T](self: Result[T, ref Exception]): T =
+  ## Fetch value of result if set, or raise the contained exception
+  if self.isErr:
+    raise self.error
+  self.value
+
+proc `$`*[T, E](self: Result[T, E]): string =
+  ## Returns string representation of `self`
+  if self.isOk: "Ok(" & $self.value & ")"
+  else: "Err(" & $self.error & ")"
+
+template derefOr[T, E](self: Result[T, E], def: T): T =
+  ## Fetch value of result if set, or supplied default
+  ## default will not be evaluated iff value is set
+  if self.isErr: def
+  else: self.value
+
+iterator items[T, E](self: Result[T, E]): T =
+  ## Iterate over result as if it were a collection of either 0 or 1 items
```

Comment:
I'm not sure about this. It makes more sense for `items` to be lifter to iterate over the contents of `T`.

Reply:
these are nice to have when composing operations in a functional style - allows you to use generic code like `any`, `map` etc that operates on collections in some ways

Reply:
Yes, what I meant is that if you have a `Result[seq[int]` value, my expectation would be that I'm iterating over the integers in the sequence.